File: HW3_GrahamAlgorithm_CPP/Stack.cpp
#include <vector>
#include <stdexcept>
#include "Stack.h"

template<typename T>
class Stack {
private:
	struct Node {
		const T value;
		Node* next;
	};
public:
	Stack() : top_(nullptr), size_(0), max_size_(MAXSIZE) {}

	explicit Stack(size_t max_size) : top_(nullptr), size_(0), max_size_(max_size) {
		// I don't like it and I don't understand the point of MaxSize.
		if (max_size > MAXSIZE) {
			throw std::invalid_argument(&"Max size cannot be bigger than "[max_size]);
		}
	}

	Stack(const Stack<T>& other) : top_(nullptr), size_(0), max_size_(other.max_size_) {
		std::vector<T> vec = other.toVector();
		for (size_t i = vec.size() - 1; i < vec.size(); --i) {
			push(vec[i]);
		}
	}

	~Stack() {
		while (top_) {
			pop();
		}
	}

	bool IsEmpty() const {
		return !top_;
	}

	size_t Size() const {
		return size_;
	}

	T NextToTop() const {
		if (size_ < 2) {
			throw std::logic_error("There is no next-to-top element in the stack!");
		}
		return top_->next->value;
	}

	void Push(T item) {
		if (size_ == max_size_) {
			throw std::logic_error("Stack is full, cannot push a new element!");
		}
		top_ = new Node{ item, top_ };
		++size_;
	}

	void Pop() {
		if (size_ == 0) {
			throw std::logic_error("Stack is empty, cannot pop an element!");
		}
		Node* next = top_->next;
		delete top_;
		top_ = next;
		--size_;
	}

	T Top() const {
		if (IsEmpty()) {
			throw std::logic_error("Stack is empty, cannot get top element!");
		}
		return top_->value;
	}

	std::vector<T> ToVector() const {
		std::vector<T> result;
		Node* it = top_;
		while (it != nullptr) {
			result.push_back(it->value);
			it = it->next;
		}
		return result;
	}

private:
	Node* top_;
	size_t max_size_;
	size_t size_;
	static const size_t MAXSIZE = 1000000;
};

File: HW2_GrahamAlgorithm/Stack.cs
using System;

namespace GrahamAlgorithm
{
    /// <summary>
    /// Stack with size restriction.
    /// </summary>
    /// <remarks> Implemented with nodes. </remarks>
    class Stack<T>
    {
        private Node top;
        private int maxSize;

        public const int MaxSize = 1000000;

        public int Size { get; private set; }

        public Stack(int maxSize)
        {
            // I don't like it and I don't understand the point of MaxSize. 
            if (maxSize > MaxSize)
            {
                throw new ArgumentException($"Max size cannot be bigger than {MaxSize}!");
            }

            top = null;
            Size = 0;
            this.maxSize = maxSize;
        }

        /// <summary>
        /// Indicates whether the stack is empty.
        /// </summary>
        public bool IsEmpty() => Size == 0;

        /// <summary>
        /// Returns the top element of the stack. 
        /// Throws exception if the stack is there is no next-to-top element.
        /// </summary>
        /// <exception cref="InvalidOperationException"></exception>
        public T NextToTop()
        {
            if (Size < 2)
            {
                throw new InvalidOperationException("There is no next-to-top element in the stack!");
            }
            return top.Next.Value;
        }

        /// <summary>
        /// Adds an element at the top of the stack. Throws exception if the stack is full.
        /// </summary>
        /// <exception cref="InvalidOperationException"></exception>
        public void Push(T item)
        {
            if (Size == maxSize)
            {
                throw new InvalidOperationException("Stack is full, cannot push a new element!");
            }
            top = new Node(item, top);
            ++Size;
        }

        /// <summary>
        /// Removes an element from top of the stack. 
        /// Throws exception if the stack is empty.
        /// </summary>
        /// <exception cref="InvalidOperationException"></exception>
        public void Pop()
        {
            if (Size == 0)
            {
                throw new InvalidOperationException("Stack is empty, cannot pop an element!");
            }
            top = top.Next;
            --Size;
        }

        /// <summary>
        /// Returns the top element of the stack. Throws exception if the stack is empty.
        /// </summary>
        /// <exception cref="InvalidOperationException"></exception>
        public T Top()
        {
            if (IsEmpty())
            {
                throw new InvalidOperationException("Stack is empty, cannot get top element!");
            }
            return top.Value;
        }

        /// <summary>
        /// Converts the stack to an array (from top to the bottom).
        /// </summary>
        public T[] ToArray()
        {
            T[] result = new T[Size];
            if (Size == 0)
            {
                return result;
            }
            Node it = top;
            for (int i = 0; i < Size; ++i)
            {
                result[i] = it.Value;
                it = it.Next;
            }
            return result;
        }


        private class Node
        {
            public T Value { get; }
            public Node Next { get; }

            public Node(T value, Node next)
            {
                Value = value;
                Next = next;
            }
        }
    }
}
